### PR TITLE
feat: add npm scope [CSENG-94]

### PR DESCRIPTION
**Pull request by snyk-abedonik**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

This PR adds support for npm scope labels on resolved dependency packages. When the `showNpmScope` option is enabled, packages in the dependency tree will be annotated with an `npm:scope` label that indicates their dependency classification:

- `dev` - for development dependencies
- `prod` - for production dependencies  
- `unknown` - for extraneous and optional dependencies

The feature is implemented through a new `dep-labels.ts` module that provides the `withPackageLabels()` function, which is integrated throughout the dependency resolution pipeline in both `deps.ts` (file tree) and `logical.ts` (logical tree) modules.

#### Where should the reviewer start?

1. **Start with the new module**: `lib/dep-labels.ts` - This contains the core logic for determining and assigning npm scope labels based on dependency types.

2. **Review type definitions**: `lib/types.ts` - Check the new `NpmScope` type, `PackageLabels` interface, and the `showNpmScope` option added to the `Options` interface.

3. **Integration points**: 
   - `lib/deps.ts` (lines 132-135, 231-234) - Where labels are added to root packages and dependencies in the file tree
   - `lib/logical.ts` (lines 85-89, 96-100, 170-174) - Where labels are added during logical tree construction, including for extraneous dependencies

4. **Test coverage**: `test/logical.test.js` (lines 154-267) - Comprehensive test suite covering all scenarios including edge cases.

#### How should this be manually tested?

1. **Test with `showNpmScope: true`**:
   ```javascript
   const resolveTree = require('./lib');
   const result = await resolveTree('./test/fixtures/bundle', { 
     showNpmScope: true, 
     dev: true 
   });
   
   // Verify all dependencies have labels
   // Verify dev dependencies have labels['npm:scope'] === 'dev'
   // Verify prod dependencies have labels['npm:scope'] === 'prod'
   // Verify extraneous and optional dependencies have labels['npm:scope'] === 'unknown'
   ```

2. **Test with `showNpmScope: false` or undefined**:
   ```javascript
   const result = await resolveTree('./test/fixtures/bundle', { 
     showNpmScope: false 
   });
   // Verify no packages have labels property
   ```

3. **Run the test suite**:
   ```bash
   npm test -- test/logical.test.js
   ```

4. **Test nested dependencies**: Verify that nested dependencies at all levels of the tree also receive appropriate labels when `showNpmScope: true`.

#### Any background context you want to provide?

This feature enables downstream consumers of the dependency tree to easily identify and filter packages based on their npm scope classification (dev vs prod). The implementation is opt-in via the `showNpmScope` option to maintain backward compatibility and avoid unnecessary overhead when labels are not needed.

The label assignment logic maps dependency types as follows:
- `DEV` → `"dev"`
- `PROD` → `"prod"`  
- `EXTRANEOUS` and `OPTIONAL` → `"unknown"`

Labels are preserved when copying package objects in the logical tree (see `copy()` function in `logical.ts`), ensuring they persist through the tree transformation process.

#### What are the relevant tickets?

https://snyksec.atlassian.net/browse/CSENG-94

#### Screenshots

N/A

#### Additional questions

None

### CODE REVIEW

File: lib/dep-labels.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    "npm:scope": scope,
+  };
+}
+
+export function getNpmScope(depType: DepType): NpmScope {
+  switch (depType) {
+    case depTypes.DEV:
+      return "dev";
+    case depTypes.PROD:
+    case depTypes.OPTIONAL:
```
Comment (gitphill) [inline lib/dep-labels.ts]: is optional always prod?

**Review by gitphill — APPROVED** [review]

File: lib/dep-labels.ts (hunk 1 above)
Comment (snyk-abedonik) [reply lib/dep-labels.ts]: We will return unknown value fir optional in v1.

Thank you!

image
